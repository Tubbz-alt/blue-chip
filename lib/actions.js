"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clearResources = exports.removeResource = exports.updateResource = exports.updateResources = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _jsonApiNormalizer = require("json-api-normalizer");

var _jsonApiNormalizer2 = _interopRequireDefault(_jsonApiNormalizer);

var _graphqlNormalizr = require("graphql-normalizr");

var _jsonapiValidator = require("jsonapi-validator");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var jsonApiValidator = new _jsonapiValidator.Validator();
var graphQLNormalizr = new _graphqlNormalizr.GraphQLNormalizr();
var graphQlNormalize = graphQLNormalizr.normalize;

var updateResources = function updateResources(payload, storeUpdater) {
  if (_isGraphQl(payload)) {
    // TODO: refactor and abstract these.  Several functions can be pulled out
    Object.entries(graphQlNormalize(payload)).forEach(function (array) {
      var _array = _slicedToArray(array, 2),
          resourceType = _array[0],
          resourcesById = _array[1];

      if (_isRedux(storeUpdater)) {
        // if it is a function asume it is Redux dispatch
        storeUpdater({
          type: "MERGE_RESOURCES",
          resourceType: resourceType,
          resourcesById: _convertToJsonApiSpec(resourceType, resourcesById)
        });
      } else if (_isMobx(storeUpdater)) {
        // if it is a function asume it is MobX resources store
        // TODO: pull out into a helper function.  Same method used in the reducer
        Object.entries(_convertToJsonApiSpec(resourceType, resourcesById)).forEach(function (_ref) {
          var _ref2 = _slicedToArray(_ref, 2),
              id = _ref2[0],
              resource = _ref2[1];

          if (!storeUpdater[resourceType]) {
            storeUpdater[resourceType] = {};
          }
          storeUpdater[resourceType][id] = resource;
        });
      }
    });
  } else {
    Object.entries((0, _jsonApiNormalizer2.default)(payload)).forEach(function (_ref3) {
      var _ref4 = _slicedToArray(_ref3, 2),
          resourceType = _ref4[0],
          resourcesById = _ref4[1];

      if (_isRedux(storeUpdater)) {
        // if it is a function asume it is Redux dispatch
        storeUpdater({
          type: "MERGE_RESOURCES",
          resourceType: resourceType,
          resourcesById: resourcesById
        });
      } else if (_isMobx(storeUpdater)) {
        // TODO: pull out into a helper function.  Same method used in the reducer
        Object.entries(resourcesById).forEach(function (_ref5) {
          var _ref6 = _slicedToArray(_ref5, 2),
              id = _ref6[0],
              resource = _ref6[1];

          if (!storeUpdater[resourceType]) {
            storeUpdater[resourceType] = {};
          }
          storeUpdater[resourceType][id] = resource;
        });
      }
    });
  }
};

var updateResource = function updateResource(_ref7, storeUpdater) {
  var id = _ref7.id,
      type = _ref7.type,
      attributes = _ref7.attributes,
      links = _ref7.links,
      relationships = _ref7.relationships;

  if (_isRedux(storeUpdater)) {
    storeUpdater({
      type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
      resourceType: type,
      id: id,
      attributes: attributes,
      links: links,
      relationships: relationships || _buildRelationships(type, attributes)
    });
  } else if (_isMobx(storeUpdater)) {
    if (!(type in storeUpdater)) {
      storeUpdater[type] = {};
    }

    storeUpdater[type][id] = {
      type: type,
      id: id,
      attributes: attributes,
      links: links,
      relationships: relationships || _buildRelationships(type, attributes)
    };
  }
};

var removeResource = function removeResource(_ref8, storeUpdater) {
  var id = _ref8.id,
      type = _ref8.type;

  if (_isRedux(storeUpdater)) {
    storeUpdater({
      type: "REMOVE_RESOURCE_BY_ID",
      resourceType: type,
      id: id
    });
  } else if (_isMobx(storeUpdater)) {
    delete storeUpdater[type][id];
  }
};

var clearResources = function clearResources(resourceTypes, storeUpdater) {
  if (_isRedux(storeUpdater)) {
    storeUpdater({
      type: "CLEAR_RESOURCES",
      resourceTypes: resourceTypes
    });
  } else if (_isMobx(storeUpdater)) {
    resourceTypes.forEach(function (resourceType) {
      storeUpdater[resourceType] = {};
    });
  }
};

var _isRedux = function _isRedux(storeUpdater) {
  return typeof storeUpdater === "function";
};

var _isMobx = function _isMobx(storeUpdater) {
  return (typeof storeUpdater === "undefined" ? "undefined" : _typeof(storeUpdater)) === "object";
};

var _isGraphQl = function _isGraphQl(payload) {
  return payload["data"] && payload["data"][0] && payload["data"][0] && "__typename" in payload["data"][0];
};

var _buildRelationships = function _buildRelationships(type, resource) {
  return Object.entries(resource).reduce(function (newObject, _ref9) {
    var _ref10 = _slicedToArray(_ref9, 2),
        key = _ref10[0],
        value = _ref10[1];

    if (value && Array.isArray(value)) {
      if (!newObject[key]) {
        newObject[key] = { data: [] };
      }

      newObject[key].data = value.map(function (id) {
        return { type: key, id: id };
      });
    }

    if (value && (typeof value === "undefined" ? "undefined" : _typeof(value)) === "object") {
      // TODO: handle hasOne and belongsTo
    }
    return newObject;
  }, {});
};

var _convertToJsonApiSpec = function _convertToJsonApiSpec(resourceType, resourcesById) {
  return Object.entries(resourcesById).reduce(function (formattedResourcesById, _ref11) {
    var _ref12 = _slicedToArray(_ref11, 2),
        id = _ref12[0],
        resource = _ref12[1];

    formattedResourcesById[id] = {
      type: resourceType,
      id: id,
      attributes: resource,
      links: null,
      relationships: _buildRelationships(resourceType, resource)
    };

    return formattedResourcesById;
  }, {});
};

exports.updateResources = updateResources;
exports.updateResource = updateResource;
exports.removeResource = removeResource;
exports.clearResources = clearResources;