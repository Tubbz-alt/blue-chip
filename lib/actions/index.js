"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Actions = undefined;

var _Actions = require("./Actions");

var _Actions2 = _interopRequireDefault(_Actions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Actions = _Actions2.default;