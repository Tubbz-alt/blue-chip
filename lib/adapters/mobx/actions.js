"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clearResources = exports.removeResource = exports.removeResources = exports.updateResource = exports.updateResources = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _helpers = require("../../actions/helpers");

var updateResources = exports.updateResources = function updateResources(mutator, resourceType, resourcesById) {
  Object.entries(resourcesById).forEach(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
        id = _ref2[0],
        resource = _ref2[1];

    if (!mutator[resourceType]) {
      mutator[resourceType] = {};
    }
    mutator[resourceType][id] = resource;
  });
};

var updateResource = exports.updateResource = function updateResource(mutator, _ref3) {
  var id = _ref3.id,
      type = _ref3.type,
      attributes = _ref3.attributes,
      links = _ref3.links,
      relationships = _ref3.relationships;

  if (!(type in mutator)) {
    mutator[type] = {};
  }

  mutator[type][id] = {
    type: type,
    id: id,
    attributes: attributes,
    links: links,
    relationships: relationships || (0, _helpers.buildRelationships)(type, attributes)
  };
};

var removeResources = exports.removeResources = function removeResources(mutator, resources) {
  resources.forEach(function (_ref4) {
    var type = _ref4.type,
        id = _ref4.id;

    delete mutator[type][id];
  });
};

var removeResource = exports.removeResource = function removeResource(mutator, _ref5) {
  var id = _ref5.id,
      type = _ref5.type;

  delete mutator[type][id];
};

var clearResources = exports.clearResources = function clearResources(mutator, resourceTypes) {
  resourceTypes.forEach(function (resourceType) {
    mutator[resourceType] = {};
  });
};