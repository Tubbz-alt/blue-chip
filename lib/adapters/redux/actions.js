"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clearResources = exports.removeResource = exports.removeResources = exports.updateResource = exports.updateResources = undefined;

var _helpers = require("../../actions/helpers");

var updateResources = exports.updateResources = function updateResources(mutator, resourceType, resourcesById) {
  mutator({ type: "UPDATE_RESOURCES", resourceType: resourceType, resourcesById: resourcesById });
};

var updateResource = exports.updateResource = function updateResource(mutator, _ref) {
  var id = _ref.id,
      type = _ref.type,
      attributes = _ref.attributes,
      links = _ref.links,
      relationships = _ref.relationships;

  mutator({
    type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
    resourceType: type,
    id: id,
    attributes: attributes,
    links: links,
    relationships: relationships || (0, _helpers.buildRelationships)(type, attributes)
  });
};

var removeResources = exports.removeResources = function removeResources(mutator, resources) {
  mutator({
    type: "REMOVE_RESOURCES_BY_ID",
    resources: resources
  });
};

var removeResource = exports.removeResource = function removeResource(mutator, _ref2) {
  var id = _ref2.id,
      type = _ref2.type;

  mutator({
    type: "REMOVE_RESOURCE_BY_ID",
    resourceType: type,
    id: id
  });
};

var clearResources = exports.clearResources = function clearResources(mutator, resourceTypes) {
  mutator({
    type: "CLEAR_RESOURCES",
    resourceTypes: resourceTypes
  });
};