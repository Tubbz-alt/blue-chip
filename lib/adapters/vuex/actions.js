"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var updateResources = exports.updateResources = function updateResources(mutator, resourceType, resourcesById) {
  mutator("UPDATE_RESOURCES", { resourceType: resourceType, resourcesById: resourcesById });
};

var updateResource = exports.updateResource = function updateResource(mutator, _ref) {
  var id = _ref.id,
      type = _ref.type,
      attributes = _ref.attributes,
      links = _ref.links,
      relationships = _ref.relationships;

  console.log("Not Implemented Yet");
};

var removeResources = exports.removeResources = function removeResources(mutator, resources) {
  console.log("Not Implemented Yet");
};

var removeResource = exports.removeResource = function removeResource(mutator, _ref2) {
  var id = _ref2.id,
      type = _ref2.type;

  console.log("Not Implemented Yet");
};

var clearResources = exports.clearResources = function clearResources(mutator, resourceTypes) {
  console.log("Not Implemented Yet");
};