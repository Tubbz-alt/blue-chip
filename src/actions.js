import jsonApiNormalize from "json-api-normalizer";
import { GraphQLNormalizr } from "graphql-normalizr";
import { Validator } from "jsonapi-validator";

const jsonApiValidator = new Validator();
const graphQLNormalizr = new GraphQLNormalizr();
const graphQlNormalize = graphQLNormalizr.normalize;

const updateResources = (payload, storeUpdater) => {
  if (_isGraphQl(payload)) {
    // TODO: refactor and abstract these.  Several functions can be pulled out
    Object.entries(graphQlNormalize(payload)).forEach(array => {
      const [resourceType, resourcesById] = array;
      if (typeof storeUpdater === "function") {
        // if it is a function asume it is Redux dispatch
        storeUpdater({
          type: "MERGE_RESOURCES",
          resourceType,
          resourcesById: _convertToJsonApiSpec(resourceType, resourcesById)
        });
      } else if (typeof storeUpdater === "object") {
        // if it is a function asume it is MobX resources store
        // TODO: pull out into a helper function.  Same method used in the reducer
        Object.entries(
          _convertToJsonApiSpec(resourceType, resourcesById)
        ).forEach(([id, resource]) => {
          if (!storeUpdater[resourceType]) {
            storeUpdater[resourceType] = {};
          }
          storeUpdater[resourceType][id] = resource;
        });
      }
    });
  } else {
    Object.entries(
      jsonApiNormalize(payload)
    ).forEach(([resourceType, resourcesById]) => {
      if (typeof storeUpdater === "function") {
        // if it is a function asume it is Redux dispatch
        storeUpdater({
          type: "MERGE_RESOURCES",
          resourceType,
          resourcesById
        });
      } else if (typeof storeUpdater === "object") {
        // TODO: pull out into a helper function.  Same method used in the reducer
        Object.entries(resourcesById).forEach(([id, resource]) => {
          if (!storeUpdater[resourceType]) {
            storeUpdater[resourceType] = {};
          }
          storeUpdater[resourceType][id] = resource;
        });
      }
    });
  }
};

const updateResource = (
  { id, type, attributes, links, relationships },
  storeUpdater
) => {
  if (typeof storeUpdater === "function") {
    storeUpdater({
      type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
      resourceType: type,
      id,
      attributes,
      links,
      relationships: relationships || _buildRelationships(type, attributes)
    });
  } else if (typeof storeUpdater === "object") {
    if (!(resourceType in storeUpdater)) {
      storeUpdater[resourceType] = {};
    }

    storeUpdater[resourceType][id] = {
      type,
      id,
      attributes,
      links,
      relationships: relationships || _buildRelationships(type, attributes)
    };
  }
};

const removeResource = ({ id, type }, dispatch) => {
  dispatch({
    type: "REMOVE_RESOURCE_BY_ID",
    resourceType: type,
    id
  });
};

const _isGraphQl = payload => {
  return (
    payload["data"] &&
    payload["data"][0] &&
    payload["data"][0] &&
    "__typename" in payload["data"][0]
  );
};

const _buildRelationships = (type, resource) => {
  return Object.entries(resource).reduce((newObject, [key, value]) => {
    if (value && Array.isArray(value)) {
      if (!newObject[key]) {
        newObject[key] = { data: [] };
      }

      newObject[key].data = value.map(id => ({ type: key, id }));
    }

    if (value && typeof value === "object") {
      // TODO: handle hasOne and belongsTo
    }
    return newObject;
  }, {});
};

const _convertToJsonApiSpec = (resourceType, resourcesById) => {
  return Object.entries(
    resourcesById
  ).reduce((formattedResourcesById, [id, resource]) => {
    formattedResourcesById[id] = {
      type: resourceType,
      id,
      attributes: resource,
      links: null,
      relationships: _buildRelationships(resourceType, resource)
    };

    return formattedResourcesById;
  }, {});
};

export { updateResources, updateResource, removeResource };
