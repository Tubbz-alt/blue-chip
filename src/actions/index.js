import updateResources from "./updateResources";
import updateResource from "./updateResource";
import removeResources from "./removeResources";
import removeResource from "./removeResource";
import clearResources from "./clearResources";

export {
  updateResources,
  updateResource,
  removeResources,
  removeResource,
  clearResources
};
