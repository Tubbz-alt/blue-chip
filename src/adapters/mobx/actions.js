export const updateResources = (mutator, resourceType, resourcesById) => {
  Object.entries(resourcesById).forEach(([id, resource]) => {
    if (!mutator[resourceType]) {
      mutator[resourceType] = {};
    }
    mutator[resourceType][id] = resource;
  });
};
