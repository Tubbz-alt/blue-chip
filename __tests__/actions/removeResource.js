import {removeResource} from "../../src/actions";

const dispatch = jest.fn().mockName("dispatch");
Object.defineProperty(dispatch, "name", {value: "dispatch"});

describe("removeResource", () => {
  describe("Redux", () => {
    test("dispatches remove resource action", () => {
      const checklist = {
        id: 1,
        type: "checklists"
      };

      const deleteAction = {
        type: "REMOVE_RESOURCE_BY_ID",
        resourceType: checklist.type,
        id: checklist.id
      };

      removeResource(checklist, dispatch);
      expect(dispatch).toBeCalledWith(deleteAction);
    });
  });

  describe("MobX", () => {
    describe("removeResource", () => {
      test("clears the store for the provided resources", () => {
        const store = {
          checklists: {1: {}, 2: {}}
        };

        const resource = {id: 1, type: "checklists"};

        removeResource(resource, store);
        expect(store).toEqual({checklists: {2: {}}});
        expect(store).toMatchSnapshot();
      });
    });
  });
});
