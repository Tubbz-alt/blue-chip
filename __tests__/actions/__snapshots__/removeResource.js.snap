// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeResource MobX removeResource clears the store for the provided resources 1`] = `
Object {
  "checklists": Object {
    "2": Object {},
  },
}
`;
